apps/devhub/utils.py

```diff
@@ -84,3 +97,203 @@ def escape_validation(validation):
+
+    def __init__(self, file_, addon=None, listed=None):
+        self.addon = addon
+        self.file = None
+
+        if isinstance(file_, FileUpload):
+            save = tasks.handle_upload_validation_result
+            validate = self.validate_upload(file_, listed)
+
+            # We're dealing with a bare file upload. Try to extract the
+            # metadata that we need to match it against a previous upload
+            # from the file itself.
+            try:
+                addon_data = parse_addon(file_, check=False)
+            except ValidationError:
+                addon_data = None
+        elif isinstance(file_, File):
+            # The listed flag for a File object should always come from
+            # the status of its owner Addon. If the caller tries to override
+            # this, something is wrong.
+            assert listed is None
+
+            save = tasks.handle_file_validation_result
+            validate = self.validate_file(file_)
+
+            self.file = file_
+            self.addon = self.file.version.addon
+            addon_data = {'guid': self.addon.id,
+                          'version': self.file.version.version}
+        else:
+            raise ValueError
+
+        if addon_data:
+            # If we have a valid file, try to find an associated Addon
+            # object, and a valid former submission to compare against.
+            try:
+                self.addon = (self.addon or
+                              Addon.with_unlisted.get(guid=addon_data['guid']))
+            except Addon.DoesNotExist:
+                pass
+
+            prev_file = self.find_previous_version(addon_data['version'])
+            if prev_file:
+                # Group both tasks so the results can be merged when
+                # the jobs complete.
+                validate = group((validate, self.validate_file(prev_file)))
+
+        # When the validation jobs complete, pass the results to the
+        # appropriate annotate/save task for the object type.
+        self.task = chain(validate, save.subtask([file_.pk]))
+
+    @staticmethod
+    def validate_file(file):
+        """Return a subtask to validate a File instance."""
+        return tasks.validate_file.subtask([file.pk])
+
+    @staticmethod
+    def validate_upload(upload, is_listed):
+        """Return a subtask to validate a FileUpload instance."""
+        assert not upload.validation
+
+        return tasks.validate_upload.subtask([upload.pk, is_listed])
+
+    def find_previous_version(self, version):
+        """Find a the most recent previous version of this add-on, prior to
+        `version`, that we can use to compare validation results."""
+
+        if not self.addon:
+            return
+
+        version = Version(version)
+        statuses = amo.STATUS_PUBLIC, amo.STATUS_LITE
+
+        # Find any version which has been approved, or any beta
+        # version which has been signed.
```

Comment:
Would it be enough to simply get the latest signed version?
`files = File.objects.filter(version__addon=self.addon, is_signed=True)`

For some context, since the landing of bug 1172035, all beta files are automatically signed. Maybe we don't want to compare with beta versions?
`files = File.objects.filter(version__addon=self.addon, is_signed=True).exclude(status=amo.STATUS_BETA)`